caffe2/perfkernels/CMakeLists.txt

```diff
@@ -13,37 +13,47 @@ set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS} ${common_srcs})
-# Currently MSVC seems to have a symbol not found error while linking (related
-# to source file order?). As a result we will currently disable the perfkernel
-# in msvc.
-if (NOT MSVC AND CAFFE2_COMPILER_SUPPORTS_AVX2_EXTENSIONS)
-  add_library(Caffe2_perfkernels_avx OBJECT ${avx_srcs})
-  add_library(Caffe2_perfkernels_avx2 OBJECT ${avx2_srcs})
-  add_dependencies(Caffe2_perfkernels_avx Caffe2_PROTO c10)
-  add_dependencies(Caffe2_perfkernels_avx2 Caffe2_PROTO c10)
+if (CAFFE2_COMPILER_SUPPORTS_AVX2_EXTENSIONS)
+  add_library(Caffe2_perfkernels_avx STATIC ${avx_srcs})
+  add_library(Caffe2_perfkernels_avx2 STATIC ${avx2_srcs})
+  add_dependencies(Caffe2_perfkernels_avx Caffe2_PROTO)
+  add_dependencies(Caffe2_perfkernels_avx2 Caffe2_PROTO)
+  target_link_libraries(Caffe2_perfkernels_avx PRIVATE c10)
+  target_link_libraries(Caffe2_perfkernels_avx2 PRIVATE c10)
   if (MSVC)
-    set_target_properties(
-        Caffe2_perfkernels_avx PROPERTIES COMPILE_FLAGS "/arch:AVX")
-    set_target_properties(
-        Caffe2_perfkernels_avx2 PROPERTIES COMPILE_FLAGS "/arch:AVX2")
-    # Currently MSVC doesn't support AVX512
+    target_compile_options(Caffe2_perfkernels_avx
+        PRIVATE "/arch:AVX")
+    target_compile_options(Caffe2_perfkernels_avx2
+        PRIVATE "/arch:AVX2")
   else()
-    set_target_properties(
-        Caffe2_perfkernels_avx PROPERTIES COMPILE_FLAGS "-mavx -mf16c")
-    set_target_properties(
-        Caffe2_perfkernels_avx2 PROPERTIES COMPILE_FLAGS "-mavx2 -mfma -mavx -mf16c")
+    target_compile_options(Caffe2_perfkernels_avx
+        PRIVATE "-mavx"
+        PRIVATE "-mf16c")
+    target_compile_options(Caffe2_perfkernels_avx2
+        PRIVATE "-mavx2"
+        PRIVATE "-mfma"
+        PRIVATE "-mavx"
+        PRIVATE "-mf16c")
   endif()
-  set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS}
-      $<TARGET_OBJECTS:Caffe2_perfkernels_avx>
-      $<TARGET_OBJECTS:Caffe2_perfkernels_avx2>)
+  caffe2_interface_library (Caffe2_perfkernels_avx Caffe2_perfkernels_avx_interface)
+  caffe2_interface_library (Caffe2_perfkernels_avx2 Caffe2_perfkernels_avx2_interface)
+  list (APPEND Caffe2_DEPENDENCY_WHOLE_LINK_LIBS "Caffe2_perfkernels_avx_interface")
+  list (APPEND Caffe2_DEPENDENCY_WHOLE_LINK_LIBS "Caffe2_perfkernels_avx2_interface")
 
   if (CAFFE2_COMPILER_SUPPORTS_AVX512_EXTENSIONS)
-      add_library(Caffe2_perfkernels_avx512 OBJECT ${avx512_srcs})
-      add_dependencies(Caffe2_perfkernels_avx512 Caffe2_PROTO c10)
-      set_target_properties(
-          Caffe2_perfkernels_avx512 PROPERTIES COMPILE_FLAGS "-mavx512f -mavx512dq -mavx512vl -mavx2 -mfma -mavx -mf16c")
-      set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS}
-          $<TARGET_OBJECTS:Caffe2_perfkernels_avx512>)
+    add_library(Caffe2_perfkernels_avx512 STATIC ${avx512_srcs})
+    add_dependencies(Caffe2_perfkernels_avx512 Caffe2_PROTO)
+    target_link_libraries(Caffe2_perfkernels_avx512 PRIVATE c10)
+    target_compile_options(Caffe2_perfkernels_avx512
+        PRIVATE "-mavx512f"
```

Comment:
@xkszltl is MSVC not going to add AVX512 in near future so safe to assume CAFFE2_COMPILER_SUPPORTS_AVX512_EXTENSIONS won't be defined for MSVC? CC @dskhudia

Reply:
The story is a little bit complex.
1. `/arch` is for auto-vectorization (and I think it also defines those _AVX_ macros) and it's not required for intrinsic-based SIMD.
2. AVX512 intrinsic is already available for VS.
3. `/arch:AVX512` is a undocumented hidden flag at this point.

cl doesn't complains about `/arch:AVX512`:
```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise>cl /arch:AVX3
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27025.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

cl : Command line warning D9002 : ignoring unknown option '/arch:AVX3'
cl : Command line error D8003 : missing source filename

C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise>cl /arch:AVX512
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27025.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

cl : Command line error D8003 : missing source filename
```

Reply:
So I still keep the pre-exist behavior in this cmake file but I cannot predict the future for VS support.

Reply:
CAFFE2_COMPILER_SUPPORTS_AVX512_EXTENSIONS is nowhere defined. And CAFFE2_COMPILER_SUPPORTS_AVX512F_EXTENSIONS  is not enough since the code is also using mavx512dq and mavx512vl therefore avx512 compiler support should be properly detected in cmake/MiscCheck.cmake.

Reply:
Thanks for pointing out the issues. I'll update this to CAFFE2_COMPILER_SUPPORTS_AVX512F_EXTENSIONS . In practice just checking avx512f should be ok because if a compiler supports avx512f it should also support avx512dq/vl (and Knights Landing/Mill are the only processors that has avx512f but without avx512dq/vl). To be very precise we could change cmake/MiscCheck.cmake to check avx512dq and avx512vl supports separately but maybe as another PR.